README.md
# Rancher::Api

[![Gem Version](https://badge.fury.io/rb/rancher-api.svg)](http://badge.fury.io/rb/rancher-api)
![Downloads](http://ruby-gem-downloads-badge.herokuapp.com/rancher-api)


[![Circle CI](https://circleci.com/gh/akurkin/rancher-api/tree/master.svg?style=svg)](https://circleci.com/gh/akurkin/rancher-api/tree/master)
[![Code Climate](https://codeclimate.com/github/akurkin/rancher-api/badges/gpa.svg)](https://codeclimate.com/github/akurkin/rancher-api)
[![Test Coverage](https://codeclimate.com/github/akurkin/rancher-api/badges/coverage.svg)](https://codeclimate.com/github/akurkin/rancher-api/coverage)


Rancher::Api is a Ruby wrapper around [Rancher](http://rancher.com/) API built with [Her](http://www.her-rb.org/)

Connect to Rancher and execute requests from your ruby scripts.
It allows you to use some of the super nice features from Rancher:

- provision VMs from different cloud providers (behind the scenes Rancher uses docker-machine, so that means you can create VMs in DigitalOcean, AWS, Rackspace and many more, check [list of supported drivers by Docker Machine](https://docs.docker.com/machine/drivers/))
- deploy your containers
- watch your containers
- scale your containers as you wish

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'rancher-api'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install rancher-api

## Rancher Version

Tested with:
gem < 0.6.0 -> Rancher v1.1.0
gem >= 0.6.0 -> Rancher v1.2.0

## Usage

Configure Rancher::Api first by providing url, access and secret keys:

### Classes

- **Project**
Top level object that represents "Environment" in Rancher UI
- **Service**
Service (combines containers from the same image)
- **Machine**
Physical docker hosts
- **Instance**
represents containers that were ever installed via Rancher. Better to query containers as nested resource, cuz there can be thousands of containers that were running before and still available to query via API. Removed containers are marked as 'removed' respectively
- **Environment**
In Rancher UI these are known as **Stack**, though in API they are **environments**. We're sticking to API resource name
- **Host**
These are hosts, with detailed information about docker installation and resources

### Setup

#### Using initializer

```ruby
require 'rancher/api'

Rancher::Api.configure do |config|
  config.url = 'http://127.0.0.1:8080/v1/'
  config.access_key = '8604A1FC8C108BAFB1E3'
  config.secret_key = '4BhuyyyAaaaaBbbbi7yaZzzAaa3y13pC6D7e569'
end
```

#### Using environment variables

IMPORTANT NOTE: Use environment's API keys. This is done for compatibility with rancher-compose to utilize same keys

    By default, the API keys under the API section are account API keys and you need to create an environment API key, which is in the Advanced Options.

You can configure `rancher-api` gem using `rancher-compose`-compatible environment variables:

- RANCHER_URL
- RANCHER_ACCESS_KEY
- RANCHER_SECRET_KEY

```ruby
Rancher::Api.setup!
```

### Querying

Now, you're able to query entities like this:

```ruby
project = Rancher::Api::Project.all.to_a
machine = Rancher::Api::Machine.find('1ph1', _project_id: project.id)
```

`rancher/api` gem uses ORM Her which hence inherently supports all of the features that Her has to offer. To get more details, review this page https://github.com/remiprev/her#fetching-data

Some of the example queries include:

```ruby
project = Rancher::Api::Project.all.to_a.first

project.machines

# exact machine name
project.machines.where(name: 'ciqa01')

# machine's name not equal to
project.machines.where(name_ne: 'qa')

# machine's name starts with
project.machines.where(name_prefix: 'qa')

# other attributes

# state not active
project.machines.where(state_ne: 'active')

# state activating
project.machines.where(state: 'activating')
```

### Creating new machines
Creating new machine using **Digital Ocean** driver:

**NOTICE**: First specify driver, so that driver_config= accessor can correctly map config on the right attribute. I.e. for 'digitalocean' config attribute is 'digitaloceanConfig'.

#### Digital Ocean

```ruby
project = Rancher::Api::Project.all.to_a.first

new_machine = project.machines.build
new_machine.driver = Rancher::Api::Machine::DIGITAL_OCEAN
new_machine.driver_config = Rancher::Api::Machine::DriverConfig.new(
    accessToken: 'xyz',
    size: '1gb',
    region: 'ams3',
    image: 'ubuntu-14-04-x64'
)

new_machine.save
```

#### Vmware Vsphere

```ruby
project = Rancher::Api::Project.all.to_a.first

new_machine = project.machines.build
new_machine.name = 'api-test'
new_machine.driver = Rancher::Api::Machine::VMWARE_VSPHERE
new_machine.driver_config = Rancher::Api::Machine::DriverConfig.new(
    boot2dockerUrl: nil,
    cpuCount: '1',
    datacenter: 'ha-dc',
    datastore: 'prod',
    diskSize: '10000',
    memorySize: '1024',
    network: 'prod',
    password: 'holamundo',
    pool: nil,
    username: 'myuser',
    vcenter: 'vcenter.happyops.com',
    vcenterPort: nil
)

new_machine.save
```


### Executing shell commands in containers

```ruby
container = Rancher::Api::Instance.find('1i382')
puts container.execute('whoami').response
puts container.execute(['bundle', 'exec', 'rake', 'db:create', 'db:migrate']).response
```

## Development

### Console
To load environment with pry run `pry -I lib -r rancher/api`

Then execute `Rancher::Api.setup!` to configure rancher credentials from environment variables and load models.

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake rspec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/akurkin/rancher-api.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.7.0] - 2016-12-15

### Changed

- environment variable `RANCHER_URL` now must be set to environment url, i.e. `http://<rancher host>/v1/projects/1a9`

## [0.6.0] - 2016-12-14

### Changed

- update `Rancher::Api::Machine` model to use collection path that has `project_id` as part of url, i.e. `/v1/projects/1a5/machines/1ph1` instead of `/v1/machines/1ph1`. It seems like Rancher 1.2 forces you to use full path instead of supplying `project_id` as parameter


## [0.5.2] - 2016-08-22

- adding models for registry and registrycredentials

## [0.5.1] - 2016-08-16

- bugfix in model helpers #handle_response when handling collections

## [0.5.0] - 2016-08-10

- added few helpers:
    - #wait_for_state - waiting (EM looping) until a desired state is reached
    - #reload - reloading of resources
    - #run - run actions (custom posts)

## [0.4.0] - 2016-07-14

- update dependencies to work with Rails 5
- update README.md

## [0.3.8] - 2016-03-15

- add Ipaddress class and association to Host class

## [0.3.7] - 2016-03-15

- fix an issue with RANCHER_URL env variable (append `/v1/` to the RANCHER_URL env variable if it exists). This way RANCHER_URL becomes truly `rancher-compose`-compatible.

## [0.3.6] - 2016-03-15
### Added

- VMware vSphere driver support
- abillity to configure Rancher::Api using `rancher-compose`-compatible environment variables



examples/create_docker_host.rb
# Objectives of this script is to:
# - check if docker host for branch exists (using labels)
# - create docker host via Rancher if it doesn't exist
#
require 'rancher/api'

BRANCH = ENV['CIRCLE_BRANCH']
DIGITAL_OCEAN_ACCESS_TOKEN = 'xxx'.freeze

Rancher::Api.configure do |config|
  config.url = "http://#{ENV['RANCHER_HOST']}/v1/"
  config.access_key = ENV['RANCHER_ACCESS_KEY']
  config.secret_key = ENV['RANCHER_SECRET_KEY']
end

project = Rancher::Api::Project.all.to_a.first
all_machines = project.machines

# 1. check if docker host exists, label branch must be equal to current branch
machine = all_machines.select { |x| x.labels['branch'] == BRANCH }.first

# 2. docker host doesn't exist, let's create one
unless machine
  machine = project.machines.build
  machine.name = 'app-' + BRANCH
  machine.driver = Rancher::Api::Machine::DIGITAL_OCEAN
  machine.driver_config = Rancher::Api::Machine::DriverConfig.new(
    accessToken: DIGITAL_OCEAN_ACCESS_TOKEN,
    size: '1gb',
    region: 'ams3',
    image: 'ubuntu-14-04-x64'
  )

  machine.labels = {
    branch: BRANCH
  }

  machine.save

  puts "CREATING NEW MACHINE: #{machine.id} - #{machine.name}"
  puts 'Set timeout to 5 minutes'

  # Wait until machine is active, on Digital Ocean claim to be 55 seconds
  #
  Timeout.timeout(300) do
    i = 45
    puts "Waiting #{i} seconds..."
    sleep i

    while machine.transitioning == 'yes'
      wait_time = 5
      i += wait_time

      puts machine.transitioningMessage
      puts "Waiting total: #{i} seconds ..."

      sleep wait_time

      machine = Rancher::Api::Machine.find(machine.id, _project_id: project.id)
    end
  end
end
